### PR TITLE
Fixed issue #1770 filed in thimble repo

### PR DESCRIPTION
**Pull request by pranay414**
@flukeout filed https://github.com/mozilla/thimble.mozilla.org/issues/1770. I have made the appropriate changes and resolved the issue. We can now see the hovering effect on FILE bar when dragging files over it.

### CODE REVIEW

**Comment by flukeout** [comment]
Left a few code comments. Please fix those up, I'll be able to test the functionality tomorrow and provide some design advice. We'll probably want to add some kind of message like "Drop files here" while it's being dragged over.

Cheers!

**Comment by pranay414** [comment]
@flukeout sure I'll fix it up

File: src/utils/DragAndDrop.js [hunk 1]
```diff
@@ -183,6 +183,13 @@ define(function (require, exports, module) {
             event.preventDefault();
 
             options.ondragleave(event);
+            // selecting FILE bar 
+            // adding functionality when dragging files away from FILE bar
+            // make call to hoverOver function to change styles
+            var sidebar = document.getElementById("sidebar");
```
Comment (humphd) [inline src/utils/DragAndDrop.js]: In this code you're mixing pure DOM js with jquery, and we might as well settle on one and stick with it.  Brackets uses jquery everywhere, so you can go ahead and use it.  Here, I'd modify your code like so:

```js
$("#sidebar").on("dragleave", function() {
    $("#sidebar").removeClass("sidebar-hover");
});
```

File: src/utils/DragAndDrop.js [hunk 2]
```diff
@@ -183,6 +183,13 @@ define(function (require, exports, module) {
...
+            sidebar.addEventListener("dragleave",function hoverOver() {
+              $("#sidebar").removeClass("sidebar-hover");;
```
Comment (humphd) [inline src/utils/DragAndDrop.js]: Single semi-colon here.

File: src/utils/DragAndDrop.js [hunk 3]
```diff
@@ -191,13 +198,27 @@ define(function (require, exports, module) {
             event.preventDefault();
 
             options.ondragover(event);
+            // selecting FILE bar 
+            // adding functionality when dragging files on FILE bar
+            // make call to hoverOver function to change styles
+            var sidebar = document.getElementById("sidebar");
+            sidebar.addEventListener("dragover",function hoverOver() {
```
Comment (humphd) [inline src/utils/DragAndDrop.js]: Same idea here, just use the jquery style.

File: src/utils/DragAndDrop.js [hunk 4]
```diff
@@ -191,13 +198,27 @@ define(function (require, exports, module) {
...
             // XXXBramble: we want to reuse this in the UploadFiles modal, so treat body differently
             if(isValidDrop(event.dataTransfer.types)) {
                 if(options.elem === window.document.body) {
                     if($(".modal.instance").length === 0) {
                         dropEffect = "copy";
+                        // selecting FILE bar 
+                        // adding functionality when dropping files on FILE bar
+                        // make call to hoverOver function to change styles
+                        var sidebar = document.getElementById("sidebar");
+                        sidebar.addEventListener("drop",function hoverOver() {
```
Comment (humphd) [inline src/utils/DragAndDrop.js]: And here too.

File: index.html [hunk 5]
```diff
@@ -0,0 +1 @@
+Hello World
```
Comment (humphd) [inline index.html]: I think this file got in by mistake maybe?

Comment (pranay414) [inline index.html]: yes this one got by mistake

File: src/extensions/default/bramble/stylesheets/sidebarTheme.css [hunk 6]
```diff
@@ -21,6 +21,12 @@ div.working-set-option-btn[style] {
     text-shadow: none;
 }
 
+/* File bar hover styles */
+.sidebar-hover {
+  background-color: #333533 !important;
```
Comment (Pomax) [inline src/extensions/default/bramble/stylesheets/sidebarTheme.css]: make sure to follow the indentation rules that you see used in the rest of the file: this file has four-space indentation

**Comment by humphd** [comment]
@pranay414 how are you coming with this?  We should move this along.  It's really close it looks like.  Also @flukeout will take it for a spin and look at styling.

File: src/extensions/default/bramble/stylesheets/sidebarTheme.css (hunk 6 above)
Comment (pranay414) [inline src/extensions/default/bramble/stylesheets/sidebarTheme.css]: sorry i didn't see that

Comment (pranay414) [inline src/extensions/default/bramble/stylesheets/sidebarTheme.css]: I have made the changes

File: index.html (hunk 5 above)
Comment (humphd) [inline index.html]: This is still here, and should be removed.

File: src/extensions/default/bramble/stylesheets/sidebarTheme.css [hunk 7]
```diff
@@ -18,12 +18,20 @@ div.working-set-option-btn[style] {
...
+    text-shadow: none !important;
+}
+
 [data-theme='light-theme'] #sidebar {
     background-color: rgba(0,0,0,.2);
     text-shadow: none;
 }
 
 
+
```
Comment (humphd) [inline src/extensions/default/bramble/stylesheets/sidebarTheme.css]: Remove this blank line.

File: src/extensions/default/bramble/stylesheets/sidebarTheme.css [hunk 8]
```diff
@@ -18,12 +18,20 @@ div.working-set-option-btn[style] {
     text-shadow: none;
 }
 
+
```
Comment (humphd) [inline src/extensions/default/bramble/stylesheets/sidebarTheme.css]: Remove this blank line.

File: src/utils/DragAndDrop.js [hunk 9]
```diff
@@ -183,6 +183,10 @@ define(function (require, exports, module) {
             event.preventDefault();
 
             options.ondragleave(event);
+            // adding functionality when dragging files away from FILE bar
```
Comment (humphd) [inline src/utils/DragAndDrop.js]: Question: why are we adding these inside the event handlers?  Why aren't we just adding it once above when we initialize all the listeners?  @flukeout when you test this, can you look at this too?  It seems wrong to me, and like it will just add multiple event handlers we don't want.

File: src/utils/DragAndDrop.js [hunk 10]
```diff
@@ -191,13 +195,21 @@ define(function (require, exports, module) {
             event.preventDefault();
 
             options.ondragover(event);
+            // adding functionality when dragging files on FILE bar
+            $("#sidebar").on("dragover", function() {
```
Comment (humphd) [inline src/utils/DragAndDrop.js]: Same thing here, should happen only once on initialization, not every time we handle the event.

File: src/utils/DragAndDrop.js [hunk 11]
```diff
@@ -191,13 +195,21 @@ define(function (require, exports, module) {
...
+            });
 
             var dropEffect =  "none";
             // XXXBramble: we want to reuse this in the UploadFiles modal, so treat body differently
             if(isValidDrop(event.dataTransfer.types)) {
                 if(options.elem === window.document.body) {
                     if($(".modal.instance").length === 0) {
                         dropEffect = "copy";
+                        // adding functionality when dropping files on FILE bar
+                        $("#sidebar").on("drop", function() {
```
Comment (humphd) [inline src/utils/DragAndDrop.js]: Same thing here.

**Comment by flukeout** [comment]
Testing now!

**Comment by flukeout** [comment]
Okay @pranay414 - I made a pull request against your branch with some updates with this. Should look like this...

hoverdrop

**Comment by pranay414** [comment]
@flukeou it looks great, do i need to make any changes to it?

**Comment by flukeout** [comment]
Hi @pranay414 - yeah, if you look in the **Conversations** tab above....

image

You'll see that @humphd has left some comments on your code that we need to address.

**Comment by pranay414** [comment]
@flukeout I get it, thanks for clearing it up for me

**Comment by humphd** [comment]
NOTE: my comments relate to the fact that we also use this DnD code for this:

dnd

File: src/htmlContent/main-view.html [hunk 12]
```diff
@@ -1,22 +1,22 @@
-<!--
```
Comment (humphd) [inline src/htmlContent/main-view.html]: All these changes are unrelated to your patch.  It looks like your editor is set to automatically remove end-of-line-whitespace.  You should disable that, and only do it manually (or get one that only does it for lines you edit).  Changing non-related lines is bad practice in open source, because it masks git blame information about when the code last changed, and who changed it.

Comment (humphd) [inline src/htmlContent/main-view.html]: NOTE: you can do `git checkout master src/htmlContent/main-view.html` to fix this.

**Comment by humphd** [comment]
@pranay414 I have some comments in here from before, did you see them?  Do they make sense?  You'll see that in https://github.com/pranay414/brackets/blob/77bd3db12c00c4dd743b30f33280a06667fb0820/src/utils/DragAndDrop.js#L163-L171 we pass in an `element` from the DOM to use as a drag-and-drop target.  So instead of using `#sidebar` you should be focused on using that element.  Also, you should only be adding/removing classes once, not adding event handlers over and over again.

Can you deal with these two issues and let me know when it's ready for review again?  If you need help, let's talk on Mattermost.

**Comment by humphd** [comment]
@pranay414 I have a big refactor patch coming soon that will cause you some rebase pain if you don't get this done in the next little bit.  We should get your code in before mine to make your life easier.

https://github.com/mozilla/brackets/pull/659 is my refactor PR.

**Comment by humphd** [comment]
@flukeout we're going to have to rethink this a bit, in light of the fact that my changes in #659 make it possible to drag things into specific folders: we don't want to hide all the file/folder names in the tree, because we still want to allow you to navigate and hover over various paths (I auto-open folders that are closed while dragging).

Could you come up with some way to do an outline and/or colour change, but not actually hide the paths?  Also, @pranay414, I really want to land my rewrite of the dnd code soon, and it's going to force you to completely rewrite your code, if you don't get to this soon.  If you can't get it done, it's OK, but it will require a rewrite.  FYI.

**Comment by flukeout** [comment]
Okay - I'm guessing that your implementation is vastly different than what's going on in this PR so does it make sense just to work on top of your PR instead?

As long as you're able to add a ".dragging-over" class or something similar to the elements, I can provide styling.

**Comment by humphd** [comment]
@flukeout and I are going to take this over, thanks @pranay414 for getting it started.

**Comment by pranay414** [comment]
@humphd if there are any projects for which i can work for 3 months, then please let me know? i wanna spend my summer working with these projects!

**Comment by Pomax** [comment]
@pranay414 there is no reason to stop working on brackets and/or thimble, but for this specific issue we've realised that what we thought we needed actually needs rethinking so @humphd and @flukeout are going to figure out what we actually need out of brackets for this now - you can't implement something if you're not sure what it has to be anymore =)
